origin-growth/src/scripts/distributeRewards.js

```diff
@@ -0,0 +1,313 @@
+const { createProviders } = require('origin-token/src/config')
+
+const db = require('../models')
+const enums = require('../enums')
+const parseArgv = require('../util/args')
+
+Logger.setLogLevel(process.env.LOG_LEVEL || 'INFO')
+const logger = Logger.create('distRewards', { showTimestamp: false })
+
+// Minimum number of block confirmations to wait for before considering
+// a reward distributed.
+const MinBlockConfirmation = 8
+const BlockMiningTimeMsec = 15000 // 15sec
+
+class TokenDistributor {
+  // Note: we can't use a constructor due to the async call to defaultAccount.
+  async init(networkId) {
+    this.networkId = networkId
+    this.token = new Token({ providers: createProviders([config.networkId]) })
+    this.supplier = await this.token.defaultAccount(networkId)
+    this.web3 = this.token.web3(networkId)
+  }
+
+  async credit(ethAddress, amount) {
+    const txnReceipt = await this.token.credit(
+      this.networkId,
+      ethAddress,
+      amount
+    )
+    logger.info('Transaction')
+    logger.info('  NetworkId:        ', this.networkId)
+    logger.info('  Amount (natural): ', amount)
+    logger.info('  Amount (tokens):  ', this.token.toTokenUnit(amount))
+    logger.info('  From:             ', this.supplier)
+    logger.info('  To:               ', ethAddress)
+    logger.info('  TxnReceipt:       ', txnReceipt)
+    return txnReceipt
+  }
+}
+
+class DistributeRewards {
+  constructor(config, distributor) {
+    this.config = config
+    this.distributor = distributor
+    this.stats = {
+      numCampaigns: 0,
+      numTxns: 0,
+      distGrandTotal: BigNumber(0)
+    }
+  }
+
+  async _distributeRewards(ethAddress, rewards) {
+    // Sum up the reward amount.
+    const total = rewards
+      .map(reward => reward.amount)
+      .reduce((a1, a2) => BigNumber(a1).plus(a2))
+
+    if (!this.config.doIt) {
+      logger.info(`Would distribute reward of ${total} to ${ethAddress}`)
+      return total
+    }
+
+    logger.info(`Distributing reward of ${total} to ${ethAddress}`)
+
+    // Mark the reward row(s) as InPayment.
+    let txn = await db.sequelize.transaction()
+    try {
+      rewards.forEach(reward => {
+        reward.update({ status: enums.GrowthRewardStatuses.InPayment })
+      })
+      await txn.commit()
+    } catch (e) {
+      await txn.rollback()
+      throw e
+    }
+
+    // Send a transaction for the total amount.
+    const txnReceipt = this.distributor.credit(ethAddress, total)
+    if (!txnReceipt.status) {
+      throw new Error('EVM reverted transaction')
+    }
+
+    // Mark the rewards as Paid.
+    txn = await db.sequelize.transaction()
+    try {
+      rewards.forEach(reward => {
+        reward.update({
+          status: enums.GrowthRewardStatuses.Paid,
+          data: Object.assign({}, reward.data, {
+            txnHash: txnReceipt.transactionHash
+          })
+        })
+      })
+      await txn.commit()
+    } catch (e) {
+      await txn.rollback()
```

Comment:
I Added a logging statement. The reward will be stuck in the 'InPayment' status. This will need some manual intervention. But should happen really really rarely, when the DB goes down right after transactions were issued.

Reply:
Yup I am also not super worried, we can probably handle these cases manually